misc/mdns/src/dns.rs

```diff
@@ -0,0 +1,371 @@
+
+//! Contains methods that handle the DNS encoding and decoding capabilities not available in the
+//! `dns_parser` library.
+
+use data_encoding;
+use libp2p_core::{Multiaddr, PeerId};
+use rand;
+use std::{borrow::Cow, cmp, error, fmt, str, time::Duration};
+use {META_QUERY_SERVICE, SERVICE_NAME};
+
+/// Decodes a `<character-string>` (as defined by RFC1035) into a `Vec` of ASCII characters.
+// TODO: better error type?
+pub fn decode_character_string(mut from: &[u8]) -> Result<Cow<[u8]>, ()> {
+    if from.is_empty() {
+        return Ok(Cow::Owned(Vec::new()));
+    }
+
+    // Remove the initial and trailing " if any.
+    if from[0] == b'"' {
+        if from.len() == 1 || from.last() != Some(&b'"') {
+            return Err(());
+        }
+        let len = from.len();
+        from = &from[1..len - 1];
+    }
+
+    // TODO: remove the backslashes if any
+    Ok(Cow::Borrowed(from))
+}
+
+/// Builds the binary representation of a DNS query to send on the network.
+pub fn build_query() -> Vec<u8> {
+    let mut out = Vec::with_capacity(33);
+
+    // Program-generated transaction ID ; unused by our implementation.
+    append_u16(&mut out, rand::random());
+
+    // Flags ; 0x0 for a regular query.
+    append_u16(&mut out, 0x0);
+
+    // Number of questions.
+    append_u16(&mut out, 0x1);
+
+    // Number of answers, authorities, and additionals.
+    append_u16(&mut out, 0x0);
+    append_u16(&mut out, 0x0);
+    append_u16(&mut out, 0x0);
+
+    // Our single question.
+    // The name.
+    append_qname(&mut out, SERVICE_NAME);
+
+    // Flags.
+    append_u16(&mut out, 0x0c);
+    append_u16(&mut out, 0x01);
+
+    // Since the output is constant, we reserve the right amount ahead of time.
+    // If this assert fails, adjust the capacity of `out` in the source code.
+    debug_assert_eq!(out.capacity(), out.len());
+    out
+}
+
+/// Builds the response to the DNS query.
+///
+/// If there are more than 2^16-1 addresses, ignores the rest.
+pub fn build_query_response(
+    id: u16,
+    peer_id: PeerId,
+    addresses: impl ExactSizeIterator<Item = Multiaddr>,
+    ttl: Duration,
+) -> Result<Vec<u8>, MdnsResponseError> {
+    // Convert the TTL into seconds.
+    let ttl = duration_to_secs(ttl);
+
+    // Add a limit to 2^16-1 addresses, as the protocol limits to this number.
+    let addresses = addresses.take(65535);
+
+    // This capacity was determined empirically and is a reasonable upper limit.
+    let mut out = Vec::with_capacity(320);
+
+    append_u16(&mut out, id);
+    // Flags ; 0x84 for an answer.
+    append_u16(&mut out, 0x8400);
+    // Number of questions, answers, authorities, additionals.
+    append_u16(&mut out, 0x0);
+    append_u16(&mut out, 0x1);
+    append_u16(&mut out, 0x0);
+    append_u16(&mut out, addresses.len() as u16);
+
+    // Our single answer.
+    // The name.
+    append_qname(&mut out, SERVICE_NAME);
+
+    // Flags.
+    append_u16(&mut out, 0x000c);
+    append_u16(&mut out, 0x8001);
```

Comment:
IIUC this sets the flush-cache bit saying that this record is the sole truth. I guess this implies that this responder owns the name and no other responder is expected to answer that query.

Reply:
Without the flush-cache-bit you would use 1 (internet class).

Reply:
Oh, ok!